### DISCUSSION TITLE
스코프 체인이 무엇인가요?

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#52] [스코프]
- 스코프 체인이 무엇인가요?
  - 코드의 출력 결과를 스코프 체인이 어떻게 동작하는 지와 연관지어 설명해 주세요.
     ```js
      function outer() {
          let outerVar = 'I am outer';
          
          function inner() {
              let innerVar = 'I am inner';
              console.log(outerVar);
              console.log(innerVar);
          }
          
          inner();
          console.log(innerVar);
      }
      outer();
      ```

**Comment by publdaze** [comment]
> - 스코프 체인이 무엇인가요?
- 코드에서 변수를 참조하고자 할 때 유효한 범위의 변수를 찾아야 함
- 변수를 참조하려고 하는 코드와 동일한 스코프에 존재하지 않으면 상위 스코프에서 찾고, 상위 스코프에서 존재하지 않으면 그 상위 스코프로 찾아나가는 과정을 거침
- 여기서 각 스코프들이 계층적으로 연결된 구조를 스코프 체인이라고 함

> - 코드의 출력 결과를 스코프 체인이 어떻게 동작하는 지와 연관지어 설명해 주세요.
> ```js
> function outer() {
>     let outerVar = 'I am outer';
>     
>     function inner() {
>         let innerVar = 'I am inner';
>         console.log(outerVar);
>         console.log(innerVar);
>     }
>     
>     inner();
>     console.log(innerVar);
> }
> outer();
> ```

- 함수가 호출되는 부분부터 보면 outer 함수가 호출이 되고 outer 함수 내부에서 inner 함수가 호출이 되는 걸 볼 수 있음.
- inner함수가 호출이 되면서 첫번째 출력에서 outerVal을 출력하고 있는데, 자신의 스코프에는 outerVar 변수가 없음 => 상위 스코프에 변수가 존재하는 지 찾아 보면 상위 스코프에서 outerVar가 선언되어 있고 값은 'I am outer'이므로 'I am outer'가 출력됨
- inner함수의 두 번째 출력은 innerVar변수를 참조하는 데 자신의 스코프에 innerVar가 선언되어 있어서 해당 값인 'I am inner'가 출력됨
- inner 함수의 호출이 끝나게 되면서 innerVar 변수의 스코프는 사라지게 됨
- 세 번째 출력에서 innerVar을 참조하지만 자신의 스코프와 상위 스코프 어디에서 유효한 innerVar 변수가 존재하지 않기 때문에 reference 에러 발생

**Comment by pipisebastian** [comment]
> 스코프 체인이 무엇인가요?

- 참조할 식별자(변수,함수)를 찾기위해, 자신의 스코프부터, 상위 스코프로 계속해서 찾아내는 것을 말합니다.

> 코드의 출력 결과를 스코프 체인이 어떻게 동작하는 지와 연관지어 설명해 주세요.

- 스코프 체인이 동작하는 원리는, 실행 컨텍스트의 렉시컬 환경(lexical environment) 속,  outerEnvironment Record 와 관련이 있습니다. 먼저 자신의 environment Record 상에서 식별자를 찾지 못할 경우, outerEnvironment Record 를 통해, 외부 컨텍스트 정보, 즉 상위 스코프로 이동합니다.
- 코드를 보면, 먼저 외부함수(outer)를 실행하고, 내부함수(inner)를 실행합니다. 내부함수에서는 먼저 outerVar를 참조하는데, 이는 내부함수의 스코프에선 없기때문에, 상위 스코프인 outer함수의 스코프로 이동하고, 외부함수의 실행컨텍스트에서 식별자를 찾아, 'I am outer'를 출력합니다.
- 두번째로 innerVar를 참조하기 위해, 내부함수의 스코프를 확인합니다. 이때 내부함수의 실행컨텍스트에서 식별자를 찾아, 'I am inner'를 출력합니다.
- 마지막으로 innerVar를 찾는데, 이때 자신의 스코프, 즉 outer함수의 실행컨텍스트에선 존재하지 않고, 그 상위인 전역 컨텍스트에서도 식별자를 찾을 수 없어 reference error가 발생하게 됩니다.
